File: src/main/java/org/anhonesteffort/btc/book/Limit.java
/*
 * Copyright (C) 2016 An Honest Effort LLC.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.anhonesteffort.btc.book;

import java.util.ArrayDeque;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Queue;

public class Limit {

  private final Queue<Order>       orderQueue = new ArrayDeque<>(12); // todo: size from constructor
  private final Map<String, Order> orderMap   = new HashMap<>();

  private final long price;
  private       long volume;

  public Limit(long price) {
    this.price  = price;
    this.volume = 0;
  }

  public long getPrice() {
    return price;
  }

  public long getVolume() {
    return volume;
  }

  public Optional<Order> peek() {
    return Optional.ofNullable(orderQueue.peek());
  }

  public void add(Order order) {
    orderMap.put(order.getOrderId(), order);
    orderQueue.add(order);
    volume += order.getSizeRemaining();
  }

  public Optional<Order> remove(String orderId) {
    Optional<Order> order = Optional.ofNullable(orderMap.remove(orderId));
    if (order.isPresent()) {
      orderQueue.remove(order.get());
      volume -= order.get().getSizeRemaining();
    }
    return order;
  }

  public Optional<Order> reduce(String orderId, long size) {
    Optional<Order> order = Optional.ofNullable(orderMap.get(orderId));
    if (order.isPresent()) {
      order.get().subtract(size, price);
      volume -= size;
      if (order.get().getSizeRemaining() <= 0) {
        orderMap.remove(orderId);
        orderQueue.remove(order.get());
      }
    }
    return order;
  }

  private long getTakeSize(Order taker) {
    if (taker instanceof MarketOrder) {
      return ((MarketOrder) taker).getSizeRemainingFor(price);
    } else {
      return taker.getSizeRemaining();
    }
  }

  private Optional<Order> takeLiquidityFromNextMaker(Order taker) {
    Optional<Order> maker = Optional.ofNullable(orderQueue.peek());
    if (maker.isPresent()) {
      long volumeRemoved = maker.get().takeSize(getTakeSize(taker));

      if (maker.get().getSizeRemaining() <= 0) {
        orderMap.remove(maker.get().getOrderId());
        orderQueue.remove();
      }

      volume -= volumeRemoved;
      taker.subtract(volumeRemoved, maker.get().getPrice());
    }
    return maker;
  }

  public List<Order> takeLiquidity(Order taker) {
    List<Order>     makers = new LinkedList<>();
    Optional<Order> maker  = null;

    while (getTakeSize(taker) > 0) {
      maker = takeLiquidityFromNextMaker(taker);
      if (maker.isPresent()) {
        makers.add(maker.get());
      } else {
        break;
      }
    }

    return makers;
  }

  public void clear() {
    orderQueue.clear();
    orderMap.clear();
    volume = 0;
  }

}

File: shared/src/main/scala/org/anhonesteffort/trading/book/Limit.scala
package org.anhonesteffort.trading.book

import org.anhonesteffort.trading.book.Orders.{MarketOrder, Order}

import scala.collection.mutable

class Limit(price: Double) {

  private val map   : mutable.Map[String, Order] = new mutable.HashMap[String, Order]()
  private val queue : mutable.Queue[Order]       = new mutable.Queue[Order]()

  private var volume : Double = 0d

  def getPrice  : Double = price
  def getVolume : Double = volume

  def peek: Option[Order] = {
    queue.headOption
  }

  def add(order: Order): Unit = {
    map.put(order.getOrderId, order)
    queue.enqueue(order)
    volume += order.getSizeRemaining
  }

  def remove(orderId: String): Option[Order] = {
    map.remove(orderId) match {
      case None        => None
      case Some(order) =>
        queue.dequeueAll(_.equals(order))
        volume -= order.getSizeRemaining
        Some(order)
    }
  }

  def reduce(orderId: String, size: Double): Option[Order] = {
    map.get(orderId) match {
      case None        => None
      case Some(order) =>
        order.subtract(size, price)
        volume -= size
        if (order.getSizeRemaining <= 0d) {
          map.remove(orderId)
          queue.dequeueAll(_.equals(order))
        }
        Some(order)
    }
  }

  private def getTakeSize(taker: Order): Double = {
    taker match {
      case market : MarketOrder => market.getSizeRemainingFor(price)
      case limit  : Order       => limit.getSizeRemaining
    }
  }

  private def takeLiquidityFromNextMaker(taker: Order, takeSize: Double): Option[Order] = {
    queue.headOption match {
      case None        => None
      case Some(maker) =>
        val volumeRemoved = maker.takeSize(takeSize)

        if (maker.getSizeRemaining <= 0d) {
          map.remove(maker.getOrderId)
          queue.dequeue()
        }

        volume -= volumeRemoved
        taker.subtract(volumeRemoved, price)
        Some(maker)
    }
  }

  def takeLiquidity(taker: Order): Seq[Order] = {
    val makers   : mutable.Buffer[Order] = new mutable.ArrayBuffer[Order]()
    var takeSize : Double                = getTakeSize(taker)

    while (takeSize > 0d) {
      takeLiquidityFromNextMaker(taker, takeSize) match {
        case None        => takeSize = -1d;
        case Some(maker) =>
          makers  += maker
          takeSize = getTakeSize(taker)
      }
    }

    makers
  }

  def clear(): Unit = {
    queue.clear()
    map.clear()
    volume = 0d
  }

}
